### PR TITLE
Fix decorator not being applied to transient dependencies

### LINKED ISSUE #940: bug: Decorator is not applied properly
**Describe the bug**
Decorator is not applied properly when it is not required through a "direct" dependency.

**To Reproduce**
See https://github.com/uber-go/fx/pull/939.

**Expected behavior**
The test added in https://github.com/uber-go/fx/pull/939 

Internal Ref: GO-1668

### PR DESCRIPTION
**Pull request by sywhang**
There is a bug that only manifests when decorating objects in Fx where transient dependency types are not properly decorated. For example, if type A depends on type B, and type B depends on type C which has a decorator, it fails to recognize that.

This only occurs when all constructors and the decorators for all these types are provided at the "root" level fx.App.

This happens because fx injects a fake "root" Scope between the actual root Container (which is where constructors are provided to by default).

Fixed by making fx stop create this fake root Scope and use the dig.Container directly.

Fixes #940.

### CODE REVIEW

File: module.go [hunk 1]
```diff
@@ -81,23 +81,32 @@ func (o moduleOption) apply(mod *module) {
...
+
 // builds the Scopes using the App's Container. Note that this happens
 // after applyModules' are called because the App's Container needs to
 // be built for any Scopes to be initialized, and applys' should be called
 // before the Container can get initialized.
 func (m *module) build(app *App, root *dig.Container) {
 	if m.parent == nil {
-		m.scope = root.Scope(m.name)
-		// TODO: Once fx.Decorate is in-place,
-		// use the root container instead of subscope.
```
Comment (abhinav) [inline module.go:100]: heh

**Review by abhinav — APPROVED** [review]

**Comment by zenthangplus** [comment]
Thank, could you publish a release?

**Comment by sywhang** [comment]
@zenthangplus Sure, I can. But can you wait until Monday, please? I'll need to verify the new version with Uber's Go Monorepo and we don't do releases on Friday to avoid causing issues over the weekend for services.

**Comment by zenthangplus** [comment]
@sywhang i've tested, fx.Decorate work well, but seems fx.Replace is not work as expected. As i known, Replace is equivalent with Decorate.

**Comment by sywhang** [comment]
> @sywhang i've tested, fx.Decorate work well, but seems fx.Replace is not work as expected. As i known, Replace is equivalent with Decorate.

Can you be more specific? What's the code you wrote to test it?

**Comment by zenthangplus** [comment]
@sywhang I've add two test cases, please take a look: https://github.com/uber-go/fx/pull/942

**Comment by dannyposi** [comment]
@sywhang could you help to release, thank

**Comment by dannyposi** [comment]
Oops, seems have some problem on master branch. When upgrading to latest code on master branch, my app is not run correctly, i don't know why, but i'm investigating.

I'm using the bellow function to handle request (for testing only). 
```
httpHandler := func(w http.ResponseWriter, r *http.Request) {
		ctx := context.Background()
		ctxWithResponseWriter := context.WithValue(ctx, "w", w)
		ctxWithRequest := context.WithValue(ctxWithResponseWriter, "r", r)
		_ = app.Start(ctxWithRequest)
	}

and append hook

lc.Append(
		fx.Hook{
			OnStart: func(ctx context.Context) error {
				log.Info("start application") // Problem: only show once time
				r := ctx.Value("r").(*http.Request)
				w := ctx.Value("w").(*httptest.ResponseRecorder)
				httpServer.Handler.ServeHTTP(w, r)
				return nil
			},
		},
	)
```

But when run a test suite with two test cases, the last test case seems not run correctly. OnStart hook seems only run once time.

**Comment by zenthangplus** [comment]
Oops, seems have some problem on master branch. When upgrading to latest code on master branch, my app is not run correctly, i don't know why, but i'm investigating.

I'm using the bellow function to handle request (for testing only). 
```
httpHandler := func(w http.ResponseWriter, r *http.Request) {
		ctx := context.Background()
		ctxWithResponseWriter := context.WithValue(ctx, "w", w)
		ctxWithRequest := context.WithValue(ctxWithResponseWriter, "r", r)
		_ = app.Start(ctxWithRequest)
	}

and append hook

lc.Append(
		fx.Hook{
			OnStart: func(ctx context.Context) error {
				log.Info("start application") // Problem: only show once time
				r := ctx.Value("r").(*http.Request)
				w := ctx.Value("w").(*httptest.ResponseRecorder)
				httpServer.Handler.ServeHTTP(w, r)
				return nil
			},
		},
	)
```

But when run a test suite with two test cases, the last test case seems not run correctly. OnStart hook seems only run once time.

Then i've tried to revert to v1.18.1 and it works correctly. Seems latest code on master has problem.

**Comment by abhinav** [comment]
Thanks for catching that @zenthangplus.
Would you mind creating a new issue for this? We nearly missed the comment.
If your investigation yields a reproduction of the bug, that would be really helpful for us to fix this.

**Comment by zenthangplus** [comment]
@abhinav i've check, it related to this MR https://github.com/uber-go/fx/pull/931, it not been released yet

**Comment by sywhang** [comment]
Hey @zenthangplus, could you please specify what you mean? OnStart/OnStop hooks are only supposed to run once. If you were relying on it being run twice, that's not the intended behavior.

**Comment by zenthangplus** [comment]
@sywhang In my case, i've wrote test for integration api testing. every http call to my api will be handled by this function.

```
httpHandler := func(w http.ResponseWriter, r *http.Request) {
		ctx := context.Background()
		ctxWithResponseWriter := context.WithValue(ctx, "w", w)
		ctxWithRequest := context.WithValue(ctxWithResponseWriter, "r", r)
		_ = app.Start(ctxWithRequest)
	}
```

Before it works well, but after this MR https://github.com/uber-go/fx/pull/931 `app.Start` is only allow to run once. I can change my logic to handle `app.Stop` before handle another request but i think this is breaking change and it break my test logic.

File: decorate_test.go [hunk 2]
```diff
@@ -313,6 +313,39 @@ func TestDecorateSuccess(t *testing.T) {
...
+		}
+		type Coffee struct {
+			Type  string
+			Name  string
+			Price int
+		}
+		type PriceService struct {
+			DefaultPrice int
+		}
+		app := fxtest.New(t,
```
Comment (jkanywhere) [inline decorate_test.go:328]: This test would pass even if none of the functions executed. What guarantees the assertions on 337 and 344 are called?

Comment (sywhang) [reply decorate_test.go:328]: Invokes are guaranteed to run (and the fact that it ran is guaranteed by `fx.RequireStart().RequireStop()` - if any Invokes fail to run, that fails the test). Invoke running successfully means assertion in line 344 are called.

And yes, line 337 in itself does not guarantee the execution, but the assertion in line 344 means both constructors `Provide`d here ran successfully.

**Pull request merged**